libs/mock-block-dock/src/datastore/use-mock-datastore/shared.ts

```diff
@@ -0,0 +1,28 @@
+import { DependencyList, useCallback } from "react";
+
+import { randomNumberFromRange } from "../../util";
+
+type AnyAsyncFunction = (...args: any[]) => Promise<any>;
+
+export const useCallbackWithLatency = <T extends AnyAsyncFunction>(
+  callback: T,
+  deps: DependencyList,
+  simulateLatency?: { min: number; max: number },
+): T => {
+  // eslint-disable-next-line react-hooks/exhaustive-deps -- it's not smart enough to deal with the cast
```

Comment:
I'm not sure it's the cast – it's got no way of knowing what `callback` is doing and what the contents of `deps` is and whether the two are related at all.

I think the bigger issue though is that `useCallbackWithLatency` removes any checking of dependencies for any function passed into it, i.e. if you go and empty out a dependency array from `updateEntity`, there is no error.

I can think of two options to deal with this:

1. We keep the existing functions defined exactly as they are now (which keeps the checking of the dependency array), and then do something like `const updateEntityWithLatency = useCallbackWithLatency(...)` which is the function that gets returned from `use-mock-datastore`

2. Rather than latency being implemented by wrapping datastore functions, it's just a function that we call inside them, e.g.
```ts
  const sleepForLatency = useCallback(async () => {
    const waitFor = randomNumberFromRange(
      simulateLatency?.min ?? 0,
      simulateLatency?.max ?? 0,
    );
    return new Promise((resolve) => {
      setTimeout(resolve, waitFor);
    });
  }, [simulateLatency]);

  const queryEntities: GraphEmbedderMessageCallbacks["queryEntities"] =
    useCallback(
      async ({ data }) => {
        await sleepForLatency();
        // etc
```

Reply:
> I'm not sure it's the cast – it's got no way of knowing what callback is doing and what the contents of deps is and whether the two are related at all.

If you remove the cast it actually does stop complaining about it and only complains about the fact that we're spreading `deps`.

With cast:
<img width="1067" alt="image" src="https://user-images.githubusercontent.com/25749103/224971725-0a49bfd4-ecdb-46a7-9c56-d8e5add5e15b.png">

Without cast:
<img width="1324" alt="image" src="https://user-images.githubusercontent.com/25749103/224971852-66e7f2d6-5452-4976-a0ce-1df7f5d86c48.png">
<img width="1097" alt="image" src="https://user-images.githubusercontent.com/25749103/224971804-13952861-1a18-4a9f-8ccd-b61e8e7b6c68.png">


I agree with the general point that it doesn't know what's inside callback, but my thought was that's fine honestly because the dependencies in _this_ function are obvious and we can ensure them ourselves without relying on eslint. Unfortunately you've now made me realise that this hook is pretty dangerous because we're not getting the `deps` checked in the callers, because the eslint rule looks for `useCallback`, so it doesn't check the parameters of `useCallbackWithLatency`.
<img width="570" alt="image" src="https://user-images.githubusercontent.com/25749103/224972620-09af6c3a-4495-4cd8-94d3-84d73931b575.png">

As such I'm going to go for the plain approach (which is what I started with) that you suggest as (2). My only concern with it is how easy it would be to miss calling `sleepForLatency`, because we've got to manually call it in every callback which feels uncomfortable.

[Call the wait function manually in every callback](https://github.com/blockprotocol/blockprotocol/pull/1161/commits/4ff8d5f2ac42a92efc59df9297fb0e727aa9a5ef)